up-streamer/src/ustreamer.rs

```diff
@@ -145,37 +190,95 @@ impl ForwardingListeners {
+        {
+            *active += 1;
+            if *active > 1 {
+                return Ok(None);
+            } else {
+                return Ok(Some(forwarding_listener.clone()));
+            }
+        }
 
-                let reg_res = task::block_on(in_transport
-                    .register_listener(&any_uuri(), Some(&uauthority_to_uuri(out_authority)), forwarding_listener.clone()));
+        let forwarding_listener =
+            Arc::new(ForwardingListener::new(forwarding_id, out_sender.clone()));
 
-                if let Err(err) = reg_res {
-                    warn!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} unable to register listener, error: {err}");
-                } else {
-                    debug!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} able to register listener");
-                }
+        type UUriPair = Vec<(UUri, Option<UUri>)>;
+        let mut uuri_to_backpedal: HashSet<UUriPair> = HashSet::new();
 
-                (
-                    0,
-                    forwarding_listener,
-                )
-            });
-        *active += 1;
+        // Perform async registration and fetching
 
-        if *active > 1 {
-            None
+        uuri_to_backpedal.insert(vec![(any_uuri(), Some(uauthority_to_uuri(out_authority)))]);
+        if let Err(err) = in_transport
+            .register_listener(
+                &any_uuri(),
+                Some(&uauthority_to_uuri(out_authority)),
+                forwarding_listener.clone(),
+            )
+            .await
+        {
+            warn!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} unable to register request listener, error: {err}");
+            for uuri_pair in &uuri_to_backpedal {
+                for (uuri1, uuri2) in uuri_pair {
+                    // Do something with uuri1 and uuri2
+                    let _ = in_transport
+                        .unregister_listener(uuri1, uuri2.as_ref(), forwarding_listener.clone())
+                        .await;
+                }
+            }
+            return Err(ForwardingListenerError::FailToRegisterNotificationRequestResponseListener);
         } else {
-            Some(forwarding_listener.clone())
+            debug!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} able to register request listener");
+        }
+
+        let subscribers = match subscription_cache
+            .lock()
+            .await
+            .fetch_cache_entry(out_authority.into())
+        {
+            Some(subscribers) => subscribers,
+            None => {
+                warn!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} no subscribers found for out_authority: {out_authority:?}");
+                HashSet::new()
+            }
+        };
+
+        for subscriber in subscribers {
+            uuri_to_backpedal.insert(vec![(subscriber.topic.clone(), None)]);
```

Comment:
Based on the suggestion of tuple above, then this'd just be:

```suggestion
            uuri_to_backpedal.insert((subscriber.topic.clone(), None));
```